core/src/main/java/io/grpc/internal/InternalSubchannel.java

```diff
@@ -286,6 +292,34 @@ void resetConnectBackoff() {
     }
   }
 
+  void shutdownTransports(Status reason) {
+    ManagedClientTransport savedActiveTransport;
+    ConnectionClientTransport savedPendingTransport;
+    try {
+      synchronized (lock) {
+        if (state.getState() == SHUTDOWN) {
+          return;
+        }
+        savedActiveTransport = activeTransport;
+        savedPendingTransport = pendingTransport;
```

Comment:
(The logic of `InternalSubchannel` is pretty involved, so I'm not 100% sure, but) It seems that:
If current state is `CONNECTING`, we should call `gotoNonErrorState(IDLE)` here. Otherwise `obtainActiveTransport()` will return null all the time and never start a new transport or return an active transport.

Reply:
I take it back. `startNewTransport()` will be called in `TransportListener.transportShutdown()` of the pending transport. So there is no problem.